packages/block-library/src/breadcrumbs/edit.js

```diff
@@ -6,52 +6,136 @@ import { InspectorControls, useBlockProps } from '@wordpress/block-editor';
+import { useEffect, useState, RawHTML } from '@wordpress/element';
 import { useServerSideRender } from '@wordpress/server-side-render';
 
 /**
  * Internal dependencies
  */
 import { useToolsPanelDropdownMenuProps } from '../utils/hooks';
+
 const separatorDefaultValue = '/';
+const typeDefaultValue = 'auto';
+
+const BREADCRUMB_TYPES = {
+	auto: {
+		help: __(
+			'Try to automatically determine the best type of breadcrumb for the template.'
+		),
+	},
+	postWithAncestors: {
+		help: __(
+			'Shows breadcrumbs based on post hierarchy. Only works for hierarchical post types.'
+		),
+		placeholderItems: [ __( 'Ancestor' ), __( 'Parent' ) ],
+	},
+	postWithTerms: {
+		help: __(
+			'Shows breadcrumbs based on taxonomy terms. Chooses the first taxonomy with assigned terms and includes ancestors if the taxonomy is hierarchical.'
+		),
+		placeholderItems: [ __( 'Category' ) ],
+	},
+};
 
 export default function BreadcrumbEdit( {
 	attributes,
 	setAttributes,
-	context: { postId, postType },
+	context: { postId, postType, templateSlug },
 } ) {
-	const { separator, showHomeLink } = attributes;
-	const isPostTypeHierarchical = useSelect(
+	const { separator, showHomeLink, type } = attributes;
+	const { post, isPostTypeHierarchical, hasTermsAssigned } = useSelect(
 		( select ) => {
 			if ( ! postType ) {
-				return null;
+				return {};
 			}
-			return select( coreStore ).getPostType( postType )?.hierarchical;
+			const _post = select( coreStore ).getEntityRecord(
+				'postType',
+				postType,
+				postId
+			);
+			const taxonomies = select( coreStore ).getTaxonomies( {
+				type: postType,
+				per_page: -1,
+			} );
+			return {
+				post: _post,
+				isPostTypeHierarchical:
+					select( coreStore ).getPostType( postType )?.hierarchical,
+				hasTermsAssigned:
+					_post &&
+					( taxonomies || [] )
+						.filter(
+							( { visibility } ) => visibility?.publicly_queryable
+						)
+						.some( ( taxonomy ) => {
+							return !! _post[ taxonomy.rest_base ]?.length;
+						} ),
+			};
 		},
-		[ postType ]
+		[ postType, postId ]
 	);
+
+	// Counter used to cache-bust `useServerSideRender`
+	//
+	// This is a catch-all signal to re-render the block when a post's title,
+	// parent ID, or terms change.
+	//
+	// This is fundamentally imperfect, because there are other entities which
+	// could change in the meantime (the titles of ancestor posts, or the
+	// labels of taxonomy terms), hence the choice to re-render systematically
+	// upon saving.
+	const [ invalidationKey, setInvalidationKey ] = useState( 0 );
+	useEffect( () => {
+		setInvalidationKey( ( c ) => c + 1 );
+	}, [ post ] );
+
 	const blockProps = useBlockProps();
 	const dropdownMenuProps = useToolsPanelDropdownMenuProps();
 	const { content } = useServerSideRender( {
 		attributes,
 		skipBlockSupportAttributes: true,
 		block: 'core/breadcrumbs',
-		urlQueryArgs: { post_id: postId },
+		urlQueryArgs: { post_id: postId, invalidationKey },
```

Comment:
I didn't care to snake-case this, since this doesn't map to anything on the server. But if you really want consistency:

```suggestion
		urlQueryArgs: { post_id: postId, invalidation_key: invalidationKey },
```

Reply:
Nah, I think it's fine :). Thank you!